AFQ/viz/utils.py

```diff
@@ -411,6 +434,615 @@ def visualize_tract_profiles(tract_profiles, scalar="dti_fa", min_fa=0.0,
+            to match pyAFQ's scale for that scalar.
+            Default: SCALE_MAT_2_PYTHON
+        """
+        self.logger = logging.getLogger('AFQ.csv')
+        self.out_folder = out_folder
+
+        if isinstance(is_mats, bool):
+            is_mats = [is_mats] * len(csv_fnames)
+
+        self.profile_dict = {}
+        for i, fname in enumerate(csv_fnames):
+            profile = pd.read_csv(fname)
+            profile['subjectID'] = \
+                profile['subjectID'].apply(
+                    lambda x: int(
+                        ''.join(c for c in x if c.isdigit())
+                    ) if isinstance(x, str) else x)
+
+            if is_mats[i]:
+                profile.rename(
+                    columns=mat_column_converter, inplace=True)
+                profile['bundle'] = \
+                    profile['bundle'].apply(
+                        lambda x: mat_bundle_converter[x])
+                for scalar, scale in mat_scale_converter.items():
+                    profile[scalar] = \
+                        profile[scalar].apply(lambda x: x * scale)
+
+            for bound, constraint in scalar_bounds.items():
+                for scalar, threshold in constraint.items():
+                    profile[scalar] = \
+                        profile[scalar].apply(
+                            lambda x: self._threshold_scalar(
+                                bound,
+                                threshold,
+                                x))
+
+            self.profile_dict[names[i]] = profile
+        if subjects is None:
+            self.subjects = self.profile_dict[names[0]]['subjectID'].unique()
+        else:
+            self.subjects = subjects
+
+    def _threshold_scalar(self, bound, threshold, val):
+        if bound == "lb":
+            if val > threshold:
+                return val
+            else:
+                return np.nan
+        elif bound == "ub":
+            if val < threshold:
+                return val
+            else:
+                return np.nan
+        else:
+            raise RuntimeError("scalar_bounds dictionary "
+                               + " formatted incorrectly. See"
+                               + " the default for reference")
+
+    def _warn_not_found(self, scalar, subject, bundle, name):
+        self.logger.warning(
+            'No scalars found for scalar ' + scalar
+            + ' for subject ' + str(subject)
+            + ' for bundle ' + bundle
+            + ' for CSV ' + name)
+
+    def _warn_nans(self, scalar, subject, bundle, name, repl_nan):
+        message = (
+            'NaNs found in scalar ' + scalar
+            + ' for subject ' + str(subject)
+            + ' for bundle ' + bundle
+            + ' for CSV ' + name)
+        if repl_nan:
+            message = message + '. These Nans were replaced with 0.'
+        self.logger.info(message)
+
+    def _warn_many_nans(self, scalar, subject, bundle, name):
```

Comment:
Wouldn't it make sense to have one logger that tells you how many nodes have nans?

Reply:
The distinction for me is that a few NaNs is sent to logger.info while many NaNs is sent to logger.warning and the profile is ignored. I could make that distinction clearer in the error message. Some profiles are all / nearly all NaNs, and we should ignore those, while some have a few NaNs at the edges, which is less serious. Most of this has to do with the profiles coming out of matlab AFQ.

Reply:
OK - that makes sense. Maybe we should make this a percentage and allow users to control that percentage? I don't like the hard-coding here. And we can still report what the percentage of nan nodes was.